### ISSUE TITLE
`view_notification` GA event being sent for Banner Notifications that are dismissed or hidden

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#6109]
## Bug Description

As discussed [here](https://github.com/google/site-kit-wp/issues/6023#issuecomment-1302675823), the `view_notification` GA event that is sent for the main authentication notification is being sent even when the notification has been dismissed, and the `view_notification` event is being sent for various components that use `BannerNotification` even when the notification is hidden due to multiple notifications being present.

## Steps to reproduce

1. Set up Site Kit with tracking enabled and use the Google Analytics Debugger extension to log GA events in the JS console.
2. Ensure that multiple notifications are present, this could include:
  ~ Use a pre-5.2 version of WordPress to show the `WPVersionBumpNotification`.
  ~ Use a site in the zero or gathering data state, or use the Tester plugin to force these states, to show `ZeroDataNotification` or `GatheringDataNotification`.
3. Navigate to `/wp-admin/admin.php?page=googlesitekit-dashboard&notification=authentication_success&slug=analytics` to show the main authentication notification, `SetupSuccessBannerNotification`.
3. Check the JS console, where `view_notification` events will be logged for the hidden notifications.
4. Dismiss the main authentication notification and reload the page with the path & query string above.
5. The `view_notification` event will be logged for the dismissed notification.

## Screenshots

_Event for main authentication notification sent although the notification has been previously dismissed:_
![image](https://user-images.githubusercontent.com/18395600/199971404-9b0dbd5b-ac1d-4834-a288-de2a8b33370e.png)

_Event for gathering data notification sent although the notification is hidden:_
![image](https://user-images.githubusercontent.com/18395600/199971142-0e7ff47e-5103-48e1-acb2-715a1be19c47.png)

## Additional Context

 - PHP Version: any
 - OS: any
 - Browser: any
 - Plugin Version: checked in 1.86.0 & 1.87.0
 - Device: any

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* All of the `view_notification` events for `BannerNotification` components should be reviewed.
* All cases where the event is sent for a dismissed or hidden notification should be addressed, so the event does not fire in these cases.

## Implementation Brief

_Note for IB author: This should probably be done in a centralised manner for example by adding an `onView` prop to `BannerNotification`, which a callback can be passed into to call `trackEvent`, and potentially using `IntersectionObserver` as a mechanism for determining when the notification is viewed._

* In `assets/js/components/notifications/BannerNotification/index.js`:
    - Add a new prop, say `onView` to the `BannerNotification` component which should be a callable function.
    - Check if the `BannerNotification` is actually visible using `IntersectionObserver`, similar to the implementation in `WidgetAreaRenderer`.
        - Add a new ref (using the `useRef` hook from `@wordpress/element`) and apply it to the root `<section>` element returned by the component.
        - Use the same ref to create an `IntersectionObserver` using the `useIntersection` ref from `react-use`.
    - Add a `useEffect` hook which will check if `onView` exists and if there is a valid "intersection" using the intersection observer. If yes, then call the `onView` function.
    - Use a local component state variable to check if the `onView` function has already been called and ensure it is called only once per page load (i.e. the function should not fire every time the component comes into view).
* Move all `trackEvent()` calls within a `useMount()`/`useEffect()` for the `view_notification` event and pass it as the `onView` prop when rendering that component's `BannerNotification` component:
    - CoreSiteBannerNotification
    - SetupSuccessBannerNotification
    - GatheringDataNotification
    - ZeroDataNotification
    - UserInputSuccessBannerNotification

### Test Coverage

 - No new tests required.

## QA Brief

* Verify that the Steps to Reproduce do not fire multiple `view_notification` events. Note that the `WPVersionBumpNotification` does not exist anymore.
* Ensure the following Banner Notifications are triggered and their `view_notification` events are only fired when these are rendered on the page:
    * SetupSuccessBannerNotification (after settings up any module or follow steps to reproduce).
    * GatheringDataNotification (follow steps to reproduce / use tester plugin)
    * ZeroDataNotification (follow steps to reproduce / use tester plugin)
    * UserInputPromptBannerNotification (NOTE: this notification will "flicker" and show momentarily before another notification hides it. This causes the view_notification event to fire "correctly". This "flicker" will be resolved in #6209.)

### Additional QA for followup PR(s).

- Check that this [defect spotted in Release QA]( https://github.com/google/site-kit-wp/issues/6564#issuecomment-1445665515) no longer occurs.
- Do a smoke/regression test on the following:
  - Lazy loading widgets on the Dashboard, ensuring they continue to stay unloaded and load at the correct times in relation to scrolling the page.
  - The Navigation Bar, ensuring that clicking on the chips navigates to the correct scroll positions, and the lazy loading widgets continue to work as expected.
- Do some further smoke testing with feature tours. Ensure there are no problems scrolling the page while the tour is active.
- Take care to test all of the above on a variety of viewports (note that feature tours only display at a 783px wide viewport and above). In particular we want to ensure things work as expected on mobile viewports where the header is not sticky vs the non-mobile viewports with a sticky header).

### QA Eng

* To test `CoreSiteBannerNotification` events, use the QAB within #4689 to create some test notifications using the Redux dev tools.

## Changelog entry

* Fix bug that caused banner notification analytics events to be sent for dismissed/hidden notifications.

**Comment by techanvil** [comment]
Thanks @eugene-manuilov, good idea. I have moved the 3rd point to a note for the IB author as suggested :+1:

**Comment by eugene-manuilov** [comment]
AC ✔️

**Comment by techanvil** [comment]
Hey @jimmymadon, the IB is looking good. A couple of details:

- This should be pretty self-evident, but it might be worth mentioning that `onView` should only be called once, for the sake of clarity.
- The relevant call to `trackEvent` in `SetupSuccessBannerNotification` is in a `useEffect`, not a `useMount`. Also, I think `UserInputSuccessBannerNotification` should be added to the list as it looks like the same problem could occur here: 
https://github.com/google/site-kit-wp/blob/36045e5aba5a8d3dc94473399dbd262e2c59e50a/assets/js/components/notifications/UserInputSuccessBannerNotification.js#L38-L40

**Comment by jimmymadon** [comment]
@techanvil Thanks for the review!

>This should be pretty self-evident, but it might be worth mentioning that onView should only be called once, for the sake of clarity.
I know we have used local component state to track if the event was fired already. However, if for some reason, a notification is actually displayed twice (because the user refreshed the page before dismissing, etc), shouldn't `trackEvent` fire twice?

>Also, I think UserInputSuccessBannerNotification should be added to the list as it looks like the same problem could occur...
Thanks - the reason I didn't spend time on checking the 2 user input notifications is because they will probably be removed as part of the Key Metrics Widget flow. But no harm in refactoring it together now for consistency.

On a separate note, all GA4 Banner and a few other notifications individually check if the `trackEvent` is fired and implement other hacks like checking ahead for dismissals. Maybe if this issue is implemented fully, another issue can be created to copy the pattern from this one to all other `view_notification` event calls?

**Comment by techanvil** [comment]
Thanks @jimmymadon.

> > This should be pretty self-evident, but it might be worth mentioning that onView should only be called once, for the sake of clarity.
> 
> I know we have used local component state to track if the event was fired already. However, if for some reason, a notification is actually displayed twice (because the user refreshed the page before dismissing, etc), shouldn't `trackEvent` fire twice?

Sorry, I should have been a bit clearer. What I mean is, `onView` should only be called once for a given banner instance, rather than say being called every time the banner comes into view.

> > Also, I think UserInputSuccessBannerNotification should be added to the list as it looks like the same problem could occur...
> 
> Thanks - the reason I didn't spend time on checking the 2 user input notifications is because they will probably be removed as part of the Key Metrics Widget flow. But no harm in refactoring it together now for consistency.

Ah, right I didn't know that. It's probably worth doing now anyway, as you say for consistency, and just to be on the safe side.

> On a separate note, all GA4 Banner and a few other notifications individually check if the `trackEvent` is fired and implement other hacks like checking ahead for dismissals. Maybe if this issue is implemented fully, another issue can be created to copy the pattern from this one to all other `view_notification` event calls?

Sounds good to me!

**Comment by bethanylang** [comment]
@jimmymadon I don't see an estimate here even though this is in EB – can you please add one? Thanks!

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

@jimmymadon 

- Tested on dev environment


Plugin config - 
- Site is in zero or gathering data state.
- User input feature flag is enabled.
- On Site Kit Dashboard -Search console and analytics zero data state notification is showing.
- User input prompt notification is hidden.

Question - I've noticed that if zero data notification is appearing and user input prompt banner is hidden. Still GA4 event sent for user input prompt notification . Is this expected ?

<details>

https://user-images.githubusercontent.com/94359491/219003408-f469404a-c388-405f-9cf0-522a6b2dfcdf.mp4

</details>

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Reviewed all the view_notification events for BannerNotification components.
- Tested all cases where the event is sent for a dismissed or hidden notification, to confirm that event does not fire in these cases.
- The view_notification event now not getting logged for the dismissed and hidden notifications.
- The issue related to user input prompt notification explained [here](https://github.com/google/site-kit-wp/issues/6109#issuecomment-1431112440) will be resolve in will be resolved in https://github.com/google/site-kit-wp/issues/6209.)
- Verified that the Steps to Reproduce do not fire multiple view_notification events.
- Tested on both main and entity dashboard.
- Tested success notification events for all modules.
- Tested notification banner events for zero and gathering data state.

<details>

![image](https://user-images.githubusercontent.com/94359491/219041907-40c37733-d3df-4b67-a777-590e1c3e31d9.png)

![image](https://user-images.githubusercontent.com/94359491/219042631-497ad448-572d-45d6-bd6c-8d549316d672.png)

</details>

**Comment by mohitwp** [comment]
Required QA:Eng as per QAB.

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on main.
- Feature tour defect reported [here](https://github.com/google/site-kit-wp/issues/6564#issuecomment-1445665515) is now resolved.
- Tested feature tour when ga4banneractivation , dashboard sharing feature flags are not enabled.
- Tested feature tour when ga4banneractivation , dashboard sharing feature flags are enabled.
- Tested Feature tour in mobile viewports.
- Tested Sticky header in mobile viewports.
- Verified Dashboard and entity dashboard chip navigation.
- Verified lazy load functionality.
- Tested all cases where the event is sent for a dismissed or hidden notification, to confirm that event does not fire in these cases.
- The view_notification event now not getting logged for the dismissed and hidden notifications.
- Verified that the Steps to Reproduce do not fire multiple view_notification events.
- Tested on both main and entity dashboard.
- Tested success notification events for all modules.
- Tested notification banner events for zero and gathering data state.

<details>


https://user-images.githubusercontent.com/94359491/221839242-86596207-81ce-464d-a683-470f80e2b9cf.mp4


https://user-images.githubusercontent.com/94359491/221839290-015d89d1-29e1-4299-b201-e152b96bd976.mp4

</details>
